{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4d887c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Instalar paquete renv paquete de R que sirve para manejar entornos reproducibles de proyectos. Su función principal es aislar y “congelar” las versiones de los paquetes que usas en un proyecto.\n",
    "#install.packages(\"renv\")\n",
    "#library(renv)\n",
    "#renv::init() Inicializar proyecto renv generando renv.lock\n",
    "# Paquetes CRAN install.packages(c(\"dplyr\", \"ggplot2\", \"viridisLite\", \"readr\", \"here\"))\n",
    "# Paquetes Bioconductor (por ejemplo para Biostrings, IRanges, GenomicRanges)\n",
    "#if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "#   install.packages(\"BiocManager\")\n",
    "#BiocManager::install(c(\"Biostrings\", \"IRanges\", \"GenomicRanges\"))\n",
    "#renv::snapshot() actualiza el lockfile con las versiones exactas usadas\n",
    "#Para restaurar en otro computador \n",
    "#library(renv)\n",
    "#renv::restore()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df9e438c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " \n",
      "\n",
      "> Will \u001b[3m\u001b[3mupdate\u001b[3m\u001b[23m 1 package.\n",
      "\n",
      "> The package (594.49 kB) is cached.\n",
      "\n",
      "\u001b[90m+ \u001b[39m\u001b[34mBiocManager\u001b[39m 1.30.16 > 1.30.\u001b[1m26\u001b[22m [bld]\n",
      "\n",
      "\u001b[36mi\u001b[39m No downloads are needed, 1 pkg (594.49 kB) is cached\n",
      "\n",
      "\u001b[32mv\u001b[39m Installed \u001b[34m\u001b[34mBiocManager\u001b[34m\u001b[39m 1.30.26  \u001b[90m\u001b[90m(37ms)\u001b[90m\u001b[39m\n",
      "\n",
      "\u001b[32mv\u001b[39m 1 pkg: upd 1 \u001b[90m\u001b[90m[1.1s]\u001b[90m\u001b[39m\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pak::pkg_install(\"BiocManager\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aeef9e54",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "Bioconductor version 3.13 (BiocManager 1.30.26), R 4.1.1 (2021-08-10)\n",
      "\n",
      "Warning message:\n",
      "\"package(s) not installed when version(s) same as or greater than current; use\n",
      "  `force = TRUE` to re-install: 'Biostrings'\"\n",
      "Old packages: 'BH', 'BiocParallel', 'Cairo', 'DBI', 'GenomeInfoDb',\n",
      "  'GenomicFeatures', 'IRdisplay', 'KernSmooth', 'MatrixGenerics', 'NMF', 'R6',\n",
      "  'RColorBrewer', 'RCurl', 'RSQLite', 'Rcpp', 'S4Vectors', 'XML', 'askpass',\n",
      "  'backports', 'bit', 'bit64', 'bitops', 'blob', 'boot', 'broom', 'caTools',\n",
      "  'cachem', 'callr', 'circlize', 'class', 'cli', 'clipr', 'cluster',\n",
      "  'codetools', 'colorspace', 'cowplot', 'cpp11', 'crayon', 'curl',\n",
      "  'data.table', 'dbplyr', 'digest', 'doParallel', 'dplyr', 'evaluate', 'fansi',\n",
      "  'farver', 'fastmap', 'filelock', 'foreach', 'foreign', 'formatR', 'fs',\n",
      "  'gargle', 'generics', 'ggalluvial', 'ggdendro', 'ggplot2', 'glue',\n",
      "  'googledrive', 'googlesheets4', 'gplots', 'gtable', 'gtools', 'haven',\n",
      "  'highr', 'htmltools', 'httr', 'isoband', 'iterators', 'jsonlite', 'knitr',\n",
      "  'labeling', 'lattice', 'lifecycle', 'lubridate', 'markdown', 'matrixStats',\n",
      "  'memoise', 'mgcv', 'mime', 'munsell', 'nlme', 'nnet', 'openssl', 'openxlsx',\n",
      "  'pbdZMQ', 'pillar', 'plyr', 'png', 'pracma', 'prettyunits', 'processx',\n",
      "  'progress', 'ps', 'purrr', 'ragg', 'readr', 'readxl', 'rematch', 'repr',\n",
      "  'reprex', 'restfulr', 'rjson', 'rlang', 'rmarkdown', 'rngtools', 'rpart',\n",
      "  'rstudioapi', 'rvest', 'scales', 'shape', 'snow', 'spatial', 'stringi',\n",
      "  'stringr', 'survival', 'svglite', 'sys', 'systemfonts', 'textshaping',\n",
      "  'tibble', 'tidyr', 'tidyselect', 'timechange', 'tinytex', 'tzdb', 'utf8',\n",
      "  'uuid', 'vctrs', 'viridisLite', 'vroom', 'withr', 'xfun', 'xml2', 'yaml',\n",
      "  'zip'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "BiocManager::install(\"Biostrings\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a23eae5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " \n",
      "\n",
      "> Will \u001b[3m\u001b[3minstall\u001b[3m\u001b[23m 11 packages.\n",
      "\n",
      "> Will \u001b[3m\u001b[3mdownload\u001b[3m\u001b[23m 11 CRAN packages (20.00 MB).\n",
      "\n",
      "\u001b[90m+ \u001b[39m\u001b[34mMetrics\u001b[39m       0.1.4  [dl] \u001b[90m(81.83 kB)\u001b[39m\n",
      "\u001b[90m+ \u001b[39m\u001b[34mR.cache\u001b[39m       0.16.0 [dl] \u001b[90m(109.67 kB)\u001b[39m\n",
      "\u001b[90m+ \u001b[39m\u001b[34mR.methodsS3\u001b[39m   1.8.2  [dl] \u001b[90m(81.17 kB)\u001b[39m\n",
      "\u001b[90m+ \u001b[39m\u001b[34mR.oo\u001b[39m          1.25.0 [dl] \u001b[90m(973.46 kB)\u001b[39m\n",
      "\u001b[90m+ \u001b[39m\u001b[34mR.utils\u001b[39m       2.12.2 [dl] \u001b[90m(1.42 MB)\u001b[39m\n",
      "\u001b[90m+ \u001b[39m\u001b[34mggthemes\u001b[39m      4.2.4  [dl] \u001b[90m(436.16 kB)\u001b[39m\n",
      "\u001b[90m+ \u001b[39m\u001b[34migraph\u001b[39m        1.4.2  [dl] \u001b[90m(8.63 MB)\u001b[39m\n",
      "\u001b[90m+ \u001b[39m\u001b[34mimmuneSIM\u001b[39m     0.8.7  [dl] \u001b[90m(4.38 MB)\u001b[39m\n",
      "\u001b[90m+ \u001b[39m\u001b[34mpoweRlaw\u001b[39m      0.70.6 [dl] \u001b[90m(3.22 MB)\u001b[39m\n",
      "\u001b[90m+ \u001b[39m\u001b[34mrepmis\u001b[39m        0.5    [dl] \u001b[90m(50.62 kB)\u001b[39m\n",
      "\u001b[90m+ \u001b[39m\u001b[34mstringdist\u001b[39m    0.9.10 [dl] \u001b[90m(609.11 kB)\u001b[39m\n",
      "\n",
      "\u001b[36mi\u001b[39m Getting 11 pkgs (20.00 MB)\n",
      "\n",
      "\u001b[32mv\u001b[39m Got \u001b[34m\u001b[34mMetrics\u001b[34m\u001b[39m 0.1.4 (x86_64-apple-darwin17.0) (81.83 kB)\n",
      "\n",
      "\u001b[32mv\u001b[39m Got \u001b[34m\u001b[34mR.methodsS3\u001b[34m\u001b[39m 1.8.2 (x86_64-apple-darwin17.0) (81.17 kB)\n",
      "\n",
      "\u001b[32mv\u001b[39m Got \u001b[34m\u001b[34mR.cache\u001b[34m\u001b[39m 0.16.0 (x86_64-apple-darwin17.0) (109.67 kB)\n",
      "\n",
      "\u001b[32mv\u001b[39m Got \u001b[34m\u001b[34mrepmis\u001b[34m\u001b[39m 0.5 (x86_64-apple-darwin17.0) (50.62 kB)\n",
      "\n",
      "\u001b[32mv\u001b[39m Got \u001b[34m\u001b[34mggthemes\u001b[34m\u001b[39m 4.2.4 (x86_64-apple-darwin17.0) (436.16 kB)\n",
      "\n",
      "\u001b[32mv\u001b[39m Got \u001b[34m\u001b[34mR.oo\u001b[34m\u001b[39m 1.25.0 (x86_64-apple-darwin17.0) (973.46 kB)\n",
      "\n",
      "\u001b[32mv\u001b[39m Got \u001b[34m\u001b[34mR.utils\u001b[34m\u001b[39m 2.12.2 (x86_64-apple-darwin17.0) (1.42 MB)\n",
      "\n",
      "\u001b[32mv\u001b[39m Got \u001b[34m\u001b[34mstringdist\u001b[34m\u001b[39m 0.9.10 (x86_64-apple-darwin17.0) (609.11 kB)\n",
      "\n",
      "\u001b[32mv\u001b[39m Got \u001b[34m\u001b[34mpoweRlaw\u001b[34m\u001b[39m 0.70.6 (x86_64-apple-darwin17.0) (3.22 MB)\n",
      "\n",
      "\u001b[32mv\u001b[39m Got \u001b[34m\u001b[34mimmuneSIM\u001b[34m\u001b[39m 0.8.7 (x86_64-apple-darwin17.0) (4.38 MB)\n",
      "\n",
      "\u001b[32mv\u001b[39m Got \u001b[34m\u001b[34migraph\u001b[34m\u001b[39m 1.4.2 (x86_64-apple-darwin17.0) (8.63 MB)\n",
      "\n",
      "\u001b[32mv\u001b[39m Installed \u001b[34m\u001b[34mMetrics\u001b[34m\u001b[39m 0.1.4  \u001b[90m\u001b[90m(71ms)\u001b[90m\u001b[39m\n",
      "\n",
      "\u001b[32mv\u001b[39m Installed \u001b[34m\u001b[34mR.cache\u001b[34m\u001b[39m 0.16.0  \u001b[90m\u001b[90m(90ms)\u001b[90m\u001b[39m\n",
      "\n",
      "\u001b[32mv\u001b[39m Installed \u001b[34m\u001b[34mR.methodsS3\u001b[34m\u001b[39m 1.8.2  \u001b[90m\u001b[90m(112ms)\u001b[90m\u001b[39m\n",
      "\n",
      "\u001b[32mv\u001b[39m Installed \u001b[34m\u001b[34mR.oo\u001b[34m\u001b[39m 1.25.0  \u001b[90m\u001b[90m(193ms)\u001b[90m\u001b[39m\n",
      "\n",
      "\u001b[32mv\u001b[39m Installed \u001b[34m\u001b[34mR.utils\u001b[34m\u001b[39m 2.12.2  \u001b[90m\u001b[90m(121ms)\u001b[90m\u001b[39m\n",
      "\n",
      "\u001b[32mv\u001b[39m Installed \u001b[34m\u001b[34mggthemes\u001b[34m\u001b[39m 4.2.4  \u001b[90m\u001b[90m(387ms)\u001b[90m\u001b[39m\n",
      "\n",
      "\u001b[32mv\u001b[39m Installed \u001b[34m\u001b[34mimmuneSIM\u001b[34m\u001b[39m 0.8.7  \u001b[90m\u001b[90m(38ms)\u001b[90m\u001b[39m\n",
      "\n",
      "\u001b[32mv\u001b[39m Installed \u001b[34m\u001b[34migraph\u001b[34m\u001b[39m 1.4.2  \u001b[90m\u001b[90m(173ms)\u001b[90m\u001b[39m\n",
      "\n",
      "\u001b[32mv\u001b[39m Installed \u001b[34m\u001b[34mpoweRlaw\u001b[34m\u001b[39m 0.70.6  \u001b[90m\u001b[90m(83ms)\u001b[90m\u001b[39m\n",
      "\n",
      "\u001b[32mv\u001b[39m Installed \u001b[34m\u001b[34mrepmis\u001b[34m\u001b[39m 0.5  \u001b[90m\u001b[90m(57ms)\u001b[90m\u001b[39m\n",
      "\n",
      "\u001b[32mv\u001b[39m Installed \u001b[34m\u001b[34mstringdist\u001b[34m\u001b[39m 0.9.10  \u001b[90m\u001b[90m(30ms)\u001b[90m\u001b[39m\n",
      "\n",
      "\u001b[32mv\u001b[39m 1 pkg + 65 deps: kept 3, added 11, dld 11 (20.00 MB) \u001b[90m\u001b[90m[6s]\u001b[90m\u001b[39m\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pak::pkg_install(\"immuneSIM\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a6dc6f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Descripción de variables:\n",
    "#name_repertoire = nombre del repertorio\n",
    "#number_of_seqs = numero de secuencias \n",
    "#species = especie\n",
    "#receptor = ig (bcr)o tr (tcr)\n",
    "#chain = cadena pesada o liviana\n",
    "#verbose = activación de mensajes en la consola durante la simulación.\n",
    "#equal_cc = distribución de recuento de clone, si todos los clonotipos tienen o no mismo tamaño\n",
    "#user_defined_alpha = uniformidad de la distribucion, valor mas alto mas desiguales los clones. \n",
    "#shm = mutaciones somaticas basadas en modelos estadisticos que definen como se distribuyen las mutaciones en las secuencias simuladas.\n",
    "#none= no se simulan mutaciones / poisson=mutaciones distribuidas aleatoriamente segun modelo poisson, sin sesgo y constante en la secuencia/data= modelo basado en perfiles reales, mutaciones se concentran en CDR, baja aleatoriedad/ naive= simula secuencias sin SHM, pero puede incluir artefactos técnicos/ motif= mutaciones dirigidas por motivos especificos, introduce sesgos basados en hotspots conocidos.\n",
    "#shm.prob = probabilidad de que ocurra el evento 15/350 son 15 eventos de mutaciones en una secuencia de 350 nucleotidos.\n",
    "#vdj_noise = ruido que se genera porque se modifica la probabilidad con la que se seleccionan ciertos genes V, D o J durante la simulación de recombinación V(D)J. va de 0-1.\n",
    "#max_cdr3_length = 100, min_cdr3_length= controlan el umbral máximo y mínimo para la longitud del aminoácido CDR3. En humanos, la longitud típica de CDR3 en IgH suele estar entre 10 y 25 aminoácidos, con una mediana alrededor de 15–16.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfb6c040",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:parallel':\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    Filter, Find, Map, Position, Reduce, anyDuplicated, append,\n",
      "    as.data.frame, basename, cbind, colnames, dirname, do.call,\n",
      "    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,\n",
      "    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,\n",
      "    pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,\n",
      "    tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    I, expand.grid, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "here() starts at /Users/catg/Desktop/SOFIAC/Gitsofia/tesisbioinf-sofia/notebooks\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "initializing sim..\n",
      "simulated sequences: 10 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Archivo guardado en: /Users/catg/Desktop/SOFIAC/Gitsofia/tesisbioinf-sofia/notebooks/../data/input/repertorio_A_insilico_10_seqs.fasta\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Cargar librerías\n",
    "\n",
    "library(Biostrings)\n",
    "library(immuneSIM)\n",
    "library(here)\n",
    "\n",
    "# Parámetro: número de secuencias\n",
    "\n",
    "n_seqs <- 10  # Cambiar a 10, 100, 1000, etc.\n",
    "\n",
    "# Simular repertorio A naive sin expansión y sin SHM en R\n",
    "\n",
    "sim_repertoire <- immuneSIM(\n",
    "  name_repertoire = paste0(\"sim_rep_\", n_seqs),\n",
    "  number_of_seqs  = n_seqs,\n",
    "  species         = \"hs\",\n",
    "  receptor        = \"ig\",\n",
    "  chain           = \"h\",\n",
    "  verbose         = TRUE,\n",
    "  equal_cc        = TRUE,\n",
    "  shm             = \"none\",\n",
    "  shm.prob        = 0,\n",
    "  vdj_noise       = 0,\n",
    "  max_cdr3_length = 15,\n",
    "  min_cdr3_length = 15\n",
    ")\n",
    "\n",
    "\n",
    "# Preparar secuencias para FASTA\n",
    "\n",
    "dna_sequences <- Biostrings::DNAStringSet(sim_repertoire$sequence)\n",
    "names(dna_sequences) <- paste0(\"seq\", seq_along(dna_sequences))\n",
    "\n",
    "\n",
    "# Crear carpeta 'data/input' si no existe\n",
    "\n",
    "# dir.create(here(\"data\", \"input\"), recursive = TRUE, showWarnings = FALSE)\n",
    "\n",
    "\n",
    "# Guardar archivo FASTA en 'data/input'\n",
    "\n",
    "out_file <- here(\"..\", \"data\", \"input\",\n",
    "                 paste0(\"repertorio_A_insilico_\", n_seqs, \"_seqs.fasta\"))\n",
    "\n",
    "writeXStringSet(dna_sequences, filepath = out_file)\n",
    "message(\"Archivo guardado en: \", out_file)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5157aa60",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/Users/catg/Desktop/SOFIA C/Gitsofia/tesisbioinf-sofia/notebooks/../data/input/repertorio_A_insilico_10_seqs.fasta'"
      ],
      "text/latex": [
       "'/Users/catg/Desktop/SOFIA C/Gitsofia/tesisbioinf-sofia/notebooks/../data/input/repertorio\\_A\\_insilico\\_10\\_seqs.fasta'"
      ],
      "text/markdown": [
       "'/Users/catg/Desktop/SOFIA C/Gitsofia/tesisbioinf-sofia/notebooks/../data/input/repertorio_A_insilico_10_seqs.fasta'"
      ],
      "text/plain": [
       "[1] \"/Users/catg/Desktop/SOFIA C/Gitsofia/tesisbioinf-sofia/notebooks/../data/input/repertorio_A_insilico_10_seqs.fasta\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "out_file <- here(\"..\", \"data\", \"input\",\n",
    "                 paste0(\"repertorio_A_insilico_\", n_seqs, \"_seqs.fasta\"))\n",
    "out_file              "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3caed2d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/Users/catg/Desktop/SOFIA C/Gitsofia/tesisbioinf-sofia/notebooks'"
      ],
      "text/latex": [
       "'/Users/catg/Desktop/SOFIA C/Gitsofia/tesisbioinf-sofia/notebooks'"
      ],
      "text/markdown": [
       "'/Users/catg/Desktop/SOFIA C/Gitsofia/tesisbioinf-sofia/notebooks'"
      ],
      "text/plain": [
       "[1] \"/Users/catg/Desktop/SOFIA C/Gitsofia/tesisbioinf-sofia/notebooks\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "here()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27d8a233",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:parallel':\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    Filter, Find, Map, Position, Reduce, anyDuplicated, append,\n",
      "    as.data.frame, basename, cbind, colnames, dirname, do.call,\n",
      "    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,\n",
      "    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,\n",
      "    pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,\n",
      "    tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    I, expand.grid, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "here() starts at /Users/catg/Desktop/SOFIAC/Gitsofia/tesisbioinf-sofia/notebooks\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "initializing sim..\n",
      "simulated sequences: 10 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Archivo guardado en: /Users/catg/Desktop/SOFIAC/Gitsofia/tesisbioinf-sofia/notebooks/../data/input/repertorio_B_insilico_10_seqs.fasta\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Cargar librerías\n",
    "\n",
    "library(Biostrings)\n",
    "library(immuneSIM)\n",
    "library(here)\n",
    "\n",
    "# Parámetro: número de secuencias\n",
    "\n",
    "n_seqs <- 10  # Cambiar a 10, 100, 1000, etc.\n",
    "\n",
    "# Simular repertorio B naive con expansión y  SHM en R\n",
    "\n",
    "sim_repertoire <- immuneSIM(\n",
    "  name_repertoire = paste0(\"sim_rep_\", n_seqs),\n",
    " number_of_seqs = 10,\n",
    "  species = \"hs\",\n",
    "  receptor = \"ig\",\n",
    "  chain = \"h\",\n",
    "  verbose= TRUE,\n",
    "  equal_cc = FALSE,\n",
    "  user_defined_alpha = 1.0,  \n",
    "  shm = \"poisson\",\n",
    "  shm.prob = 05/350,\n",
    "  vdj_noise = 0.1,\n",
    "  max_cdr3_length =18, \n",
    "  min_cdr3_length =15,\n",
    " \n",
    ")\n",
    "\n",
    "\n",
    "# Preparar secuencias para FASTA\n",
    "\n",
    "dna_sequences <- Biostrings::DNAStringSet(sim_repertoire$sequence)\n",
    "names(dna_sequences) <- paste0(\"seq\", seq_along(dna_sequences))\n",
    "\n",
    "\n",
    "# Crear carpeta 'data/input' si no existe\n",
    "\n",
    "# dir.create(here(\"data\", \"input\"), recursive = TRUE, showWarnings = FALSE)\n",
    "\n",
    "\n",
    "# Guardar archivo FASTA en 'data/input'\n",
    "\n",
    "out_file <- here(\"..\", \"data\", \"input\",\n",
    "                 paste0(\"repertorio_B_insilico_\", n_seqs, \"_seqs.fasta\"))\n",
    "\n",
    "writeXStringSet(dna_sequences, filepath = out_file)\n",
    "message(\"Archivo guardado en: \", out_file)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cde6c29",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:parallel':\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    Filter, Find, Map, Position, Reduce, anyDuplicated, append,\n",
      "    as.data.frame, basename, cbind, colnames, dirname, do.call,\n",
      "    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,\n",
      "    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,\n",
      "    pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,\n",
      "    tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    I, expand.grid, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "\n",
      "Attaching package: 'Biostrings'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "here() starts at /Users/catg/Desktop/SOFIAC/Gitsofia/tesisbioinf-sofia/notebooks\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "initializing sim..\n",
      "simulated sequences: 10 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Archivo guardado en: /Users/catg/Desktop/SOFIAC/Gitsofia/tesisbioinf-sofia/notebooks/../data/input/repertorio_C_insilico_10_seqs.fasta\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Cargar librerías\n",
    "\n",
    "library(Biostrings)\n",
    "library(immuneSIM)\n",
    "library(here)\n",
    "\n",
    "# Parámetro: número de secuencias\n",
    "\n",
    "n_seqs <- 10  # Cambiar a 10, 100, 1000, etc.\n",
    "\n",
    "# Simular repertorio B naive con expansión y  SHM en R\n",
    "\n",
    "sim_repertoire <- immuneSIM(\n",
    "  name_repertoire = paste0(\"sim_rep_\", n_seqs),\n",
    " number_of_seqs = 10,\n",
    "  species = \"hs\",\n",
    "  receptor = \"ig\",\n",
    "  chain = \"h\",\n",
    "  verbose= TRUE,\n",
    "  equal_cc = FALSE,\n",
    "  user_defined_alpha = 2.3,  \n",
    "  shm = \"data\",\n",
    "  shm.prob = 15/350,\n",
    "  vdj_noise = 0.25,\n",
    "  max_cdr3_length =22, \n",
    "  min_cdr3_length =15,\n",
    " \n",
    ")\n",
    "\n",
    "# Preparar secuencias para FASTA\n",
    "\n",
    "dna_sequences <- Biostrings::DNAStringSet(sim_repertoire$sequence)\n",
    "names(dna_sequences) <- paste0(\"seq\", seq_along(dna_sequences))\n",
    "\n",
    "\n",
    "# Crear carpeta 'data/input' si no existe\n",
    "\n",
    "# dir.create(here(\"data\", \"input\"), recursive = TRUE, showWarnings = FALSE)\n",
    "\n",
    "\n",
    "# Guardar archivo FASTA en 'data/input'\n",
    "\n",
    "out_file <- here(\"..\", \"data\", \"input\",\n",
    "                 paste0(\"repertorio_C_insilico_\", n_seqs, \"_seqs.fasta\"))\n",
    "\n",
    "writeXStringSet(dna_sequences, filepath = out_file)\n",
    "message(\"Archivo guardado en: \", out_file)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "279f45b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "'mamba' is running as a subprocess and can't modify the parent shell.\n",
      "Thus you must initialize your shell before using activate and deactivate.\n",
      "\n",
      "To initialize the current bash shell, run:\n",
      "    $ eval \"$(mamba shell hook --shell bash)\"\n",
      "and then activate or deactivate with:\n",
      "    $ mamba activate\n",
      "To automatically initialize all future (bash) shells, run:\n",
      "    $ mamba shell init --shell bash --root-prefix=~/.local/share/mamba\n",
      "If your shell was already initialized, reinitialize your shell with:\n",
      "    $ mamba shell reinit --shell bash\n",
      "Otherwise, this may be an issue. In the meantime you can run commands. See:\n",
      "    $ mamba run --help\n",
      "\n",
      "Supported shells are {bash, zsh, csh, posix, xonsh, cmd.exe, powershell, fish, nu}.\n",
      "\u001b[1m\u001b[41mcritical libmamba\u001b[m Shell not initialized\n",
      "\n",
      "'mamba' is running as a subprocess and can't modify the parent shell.\n",
      "Thus you must initialize your shell before using activate and deactivate.\n",
      "\n",
      "To initialize the current bash shell, run:\n",
      "    $ eval \"$(mamba shell hook --shell bash)\"\n",
      "and then activate or deactivate with:\n",
      "    $ mamba activate\n",
      "To automatically initialize all future (bash) shells, run:\n",
      "    $ mamba shell init --shell bash --root-prefix=~/.local/share/mamba\n",
      "If your shell was already initialized, reinitialize your shell with:\n",
      "    $ mamba shell reinit --shell bash\n",
      "Otherwise, this may be an issue. In the meantime you can run commands. See:\n",
      "    $ mamba run --help\n",
      "\n",
      "Supported shells are {bash, zsh, csh, posix, xonsh, cmd.exe, powershell, fish, nu}.\n",
      "\u001b[1m\u001b[41mcritical libmamba\u001b[m Shell not initialized\n",
      "bioconda/noarch                                             Using cache\n",
      "bioconda/osx-64                                             Using cache\n",
      "conda-forge/noarch                                          Using cache\n",
      "conda-forge/osx-64                                          Using cache\n",
      "\u001b[?25l\u001b[2K\u001b[0G\u001b[?25h\u001b[?25l\u001b[2K\u001b[0G\u001b[?25h^C\n"
     ]
    }
   ],
   "source": [
    "#Codigo llamado a asignación y clustering\n",
    "!cd ../scripts/ && sh testscript.sh "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf4f3770",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Luego de clustering archivo _clone-pass.tsv se realiza analisis de metricas de expansión clonal\n",
    "# Librerías\n",
    "library(readr)\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(viridisLite)\n",
    "library(here)  # para rutas relativas\n",
    "\n",
    "\n",
    "# Configuración de archivo\n",
    "\n",
    "carpeta_archivo <- \"output\"  \n",
    "nombre_archivo <- \"repertorio_A_insilico_10_db-pass_clone-pass.tsv\"\n",
    "\n",
    "# Construir ruta relativa\n",
    "archivo_clones <- here(carpeta_archivo, nombre_archivo)\n",
    "\n",
    "# Cargar archivo\n",
    "clones <- read_tsv(archivo_clones)\n",
    "glimpse(clones)\n",
    "\n",
    "\n",
    "# Tamaño de cada clon\n",
    "\n",
    "clon_size <- clones %>%\n",
    "  group_by(clone_id) %>%\n",
    "  summarise(n_seq = n(), .groups = \"drop\") %>%\n",
    "  arrange(desc(n_seq)) %>%\n",
    "  mutate(freq = n_seq / sum(n_seq))\n",
    "\n",
    "\n",
    "# Top 10 clones\n",
    "\n",
    "top10 <- clon_size %>%\n",
    "  slice_max(order_by = n_seq, n = 10)\n",
    "\n",
    "ggplot(top10, aes(x = reorder(clone_id, -n_seq), y = n_seq, fill = clone_id)) +\n",
    "  geom_col() +\n",
    "  scale_fill_viridis_d(option = \"plasma\") +\n",
    "  labs(title = \"Top 10 Clones VDJ\", x = \"Clone ID\", y = \"Número de secuencias\") +\n",
    "  theme_minimal() +\n",
    "  theme(legend.position = \"none\")\n",
    "\n",
    "# Curva de abundancia clonal\n",
    "\n",
    "clon_size %>%\n",
    "  mutate(rank = row_number()) %>%\n",
    "  ggplot(aes(x = rank, y = freq)) +\n",
    "  geom_line() +\n",
    "  scale_y_log10() +\n",
    "  labs(title = \"Curva de abundancia clonal\", x = \"Rango del clon\", y = \"Frecuencia (log10)\") +\n",
    "  theme_minimal()\n",
    "\n",
    "\n",
    "# Distribución acumulada\n",
    "\n",
    "clon_size %>%\n",
    "  mutate(cum_freq = cumsum(freq), rank = row_number()) %>%\n",
    "  ggplot(aes(x = rank, y = cum_freq)) +\n",
    "  geom_line() +\n",
    "  labs(title = \"Distribución acumulada de clones\", x = \"Número de clones (ordenados)\", y = \"Frecuencia acumulada\") +\n",
    "  theme_minimal()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b323f1a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Métricas de diversidad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0ff04fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Métricas de shm"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "immunesim_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
